src/libcore/ffi.rs

```diff
@@ -53,80 +52,158 @@ impl fmt::Debug for c_void {
+
+    impl<'a, 'f: 'a> Deref for VaList<'a, 'f> {
+        type Target = VaListImpl<'f>;
+
+        #[inline]
+        fn deref(&self) -> &VaListImpl<'f> {
+            unsafe { mem::transmute(self) }
+        }
+    }
+
+    impl<'a, 'f: 'a> DerefMut for VaList<'a, 'f> {
+        #[inline]
+        fn deref_mut(&mut self) -> &mut VaListImpl<'f> {
+            unsafe { mem::transmute(self) }
+        }
+    }
 }
 
-/// x86_64 ABI implementation of a `va_list`.
-#[cfg(all(target_arch = "x86_64", not(windows)))]
-#[repr(C)]
-#[derive(Debug)]
+#[cfg(all(any(target_arch = "aarch64", target_arch = "powerpc",
+              target_arch = "x86_64"),
+          any(not(target_arch = "aarch64"), not(target_os = "ios")),
+          not(windows)))]
 #[unstable(feature = "c_variadic",
            reason = "the `c_variadic` feature has not been properly tested on \
                      all supported platforms",
            issue = "44930")]
-struct VaListImpl {
-    gp_offset: i32,
-    fp_offset: i32,
-    overflow_arg_area: *mut c_void,
-    reg_save_area: *mut c_void,
+mod impls {
+    use crate::marker::PhantomData;
+    use crate::ops::{Deref, DerefMut};
+    use super::c_void;
+
+    /// AArch64 ABI implementation of a `va_list`. See the
+    /// [AArch64 Procedure Call Standard] for more details.
+    ///
+    /// [AArch64 Procedure Call Standard]:
+    /// http://infocenter.arm.com/help/topic/com.arm.doc.ihi0055b/IHI0055B_aapcs64.pdf
+    #[cfg(all(target_arch = "aarch64", not(target_os = "ios"), not(windows)))]
+    #[repr(C)]
+    #[derive(Debug)]
+    #[lang = "va_list"]
+    pub struct VaListImpl<'f> {
+        stack: *mut c_void,
+        gr_top: *mut c_void,
+        vr_top: *mut c_void,
+        gr_offs: i32,
+        vr_offs: i32,
+        _marker: PhantomData<&'f c_void>,
+    }
+
+    /// PowerPC ABI implementation of a `va_list`.
+    #[cfg(all(target_arch = "powerpc", not(windows)))]
+    #[repr(C)]
+    #[derive(Debug)]
+    #[lang = "va_list"]
+    pub struct VaListImpl<'f> {
+        gpr: u8,
+        fpr: u8,
+        reserved: u16,
+        overflow_arg_area: *mut c_void,
+        reg_save_area: *mut c_void,
+        _marker: PhantomData<&'f c_void>,
+    }
+
+    /// x86_64 ABI implementation of a `va_list`.
+    #[cfg(all(target_arch = "x86_64", not(windows)))]
+    #[repr(C)]
+    #[derive(Debug)]
+    #[lang = "va_list"]
+    pub struct VaListImpl<'f> {
+        gp_offset: i32,
+        fp_offset: i32,
+        overflow_arg_area: *mut c_void,
+        reg_save_area: *mut c_void,
+        _marker: PhantomData<&'f c_void>,
+    }
+
+    impl<'f> VaListImpl<'f> {
+        /// Convert a `VaListImpl` into a `VaList` that is binary-compatible with C's `va_list`.
+        #[inline]
+        pub fn as_va_list<'s>(&'s mut self) -> VaList<'s, 'f> {
+            VaList(self)
+        }
+    }
+
+    /// A wrapper for a `va_list`
+    #[repr(transparent)]
+    #[derive(Debug)]
+    pub struct VaList<'a, 'f: 'a>(&'a mut VaListImpl<'f>);
```

Comment:
Here's a neat trick: if you make this use named fields instead, and fix the other `VaList` definition to contain a `VaListImpl` instead of `*mut c_void`, then you can share the `Deref` and `DerefMut` impls by doing `&self.inner`, and `&mut self.inner`, respectively (the extra level of indirection in some cases is going to be stripped by deref coercions).

Oh and if you add the `PhantomData<&'a ()>` to here too (even though it's unnecessary), you can move the `VaList` definition out (and keep the `#[cfg]` on the `inner` field, or w/e name it will have).

Reply:
I'll try, but it might not work: in one case `VaList` contains a `VaListImpl`, in the other a `&'a mut VaListImpl`.